File: src/ruby/ruby-tokengenerator/token_generator.rb
require 'jwt'

class TokenGenerator
  def initialize(hmac_alg = "HS256")
    @_hmac_alg = hmac_alg
  end

  def create_token(token_id, token_string, stream_name, allowed_origins = nil, allowed_ip_addresses = nil, tracking = nil, expires_in = 60)
    allowed_origins ||= []
    allowed_ip_addresses ||= []

    header = {
      typ: 'JWT'
    }

    payload = {
      streaming: {
        tokenId: token_id,
        tokenType: 'Subscribe',
        streamName: stream_name,
        allowedOrigins: allowed_origins,
        allowedIpAddresses: allowed_ip_addresses,
        tracking: tracking
      },
      exp: Time.now.to_i + expires_in
    }

    JWT.encode(payload, token_string, algorithm=@_hmac_alg, header_fields=header)
  end
end

File: src/nodejs/node-tokengenerator/TokenGenerator.mjs
import JsonWebToken from 'jsonwebtoken';

const defaultExpiresIn = 60;

export class Limits {
  static get customViewerData() {
    return 1024;
  }
}

export default class TokenGenerator {
  /**
   *
   * @param {string} [hmacAlg = 'HS256']
   */
  constructor(hmacAlg = 'HS256') {
    this.hmacAlg = hmacAlg;
  }

  /**
   *
   * @param {number} tokenId
   * @param {string} token
   * @param {string} streamName
   * @param {string[]=} allowedOrigins
   * @param {string[]=} allowedIpAddresses
   * @param {?Tracking} tracking
   * @param {?number} [expiresIn = 60]
   * @param {?string} customViewerData - Viewer data associated with connections using this token. Max length: 1024
   * @param {?string} originCluster
   * @returns {string}
   */
  createToken(tokenId, token, streamName,
              allowedOrigins, allowedIpAddresses,
              tracking = null,
              expiresIn = defaultExpiresIn,
              customViewerData = null,
              originCluster = null) {
    const streaming = {
      tokenId: tokenId,
      tokenType: 'Subscribe',
      streamName: streamName,
      originCluster: originCluster ?? null,
      tracking: tracking ?? null,
      customViewerData: customViewerData ?? null,
      allowedOrigins: allowedOrigins ?? [],
      allowedIpAddresses: allowedIpAddresses ?? []
    };
    const payload = {
      streaming: Object.fromEntries(Object.entries(streaming).filter(([_, v]) => v !== null))
    };
    this._validateStreamingPayload(payload.streaming);

    const signOptions = {
      algorithm: this.hmacAlg,
      expiresIn: expiresIn ?? defaultExpiresIn
    };

    return JsonWebToken.sign(payload, token, signOptions);
  }

  _validateStreamingPayload(streaming) {
    if (streaming.customViewerData && streaming.customViewerData.length > Limits.customViewerData) {
      throw new Error(`customViewerData cannot be longer than: ${Limits.customViewerData}`)
    }
  }
}
